src/components/file-upload/_file-upload.scss

```diff
@@ -7,13 +7,66 @@
 @import "../label/label";
 
 @include govuk-exports("govuk/component/file-upload") {
+  $spacing-unit: govuk-spacing(1);
+
   .govuk-file-upload {
     @include govuk-font($size: 19);
     @include govuk-text-colour;
     @include govuk-focusable;
+    padding-top: $spacing-unit;
+    padding-bottom: $spacing-unit;
+
+    &:focus {
+      // "Yank" the padding with negative margin to avoid a jump
+      // when element is focused
+      margin-right: -$spacing-unit;
+      margin-left: -$spacing-unit;
+      padding-right: $spacing-unit;
+      padding-left: $spacing-unit;
+      // Use `box-shadow` to add border instead of changing `border-width`
+      // (which changes element size) and since `outline` is already used for the
+      // yellow focus state.
+      box-shadow: inset 0 0 0 4px $govuk-input-border-colour;
+
+      @include govuk-if-ie8 {
+        // IE8 doesn't support `box-shadow` so add an actual border
+        border: 4px solid $govuk-input-border-colour;
+      }
+    }
+
+    // Set "focus-within" to fix https://bugzilla.mozilla.org/show_bug.cgi?id=1430196
+    // so that component receives focus in Firefox.
+    // This can't be set together with `:focus` as all versions of IE fail
+    // to recognise `focus-within` and don't set any styles from the block
+    // when it's a selector.
+    &:focus-within {
+      margin-right: -$spacing-unit;
+      margin-left: -$spacing-unit;
+      padding-right: $spacing-unit;
+      padding-left: $spacing-unit;
+
+      // Emulate `govuk-focusable` mixin styles for this component in FF
+      outline: $govuk-focus-width solid $govuk-focus-colour;
+      outline-offset: 0;
+
+      box-shadow: inset 0 0 0 4px $govuk-input-border-colour;
+    }
   }
 
   .govuk-file-upload--error {
+    // As `upload--error` has border, it needs to have the same padding as
+    // the standard focused element.
+    margin-right: -$spacing-unit;
+    margin-left: -$spacing-unit;
+    padding-right: $spacing-unit;
+    padding-left: $spacing-unit;
     border: $govuk-border-width-form-element-error solid $govuk-error-colour;
+
+    &:focus {
+      border-color: $govuk-input-border-colour;
```

Comment:
I don't think this is working quite as intended in firefox, I suspect because of the `:focus-within` hack.

In Chrome the file input, in error state, when focused, looks like this:

![govuk-frontend-review-pr-1312 herokuapp com_components_file-upload_with-error-message_preview (3)](https://user-images.githubusercontent.com/121939/57125767-7d7c8780-6d82-11e9-822c-1ecf814e5686.png)

This is consistent with the other form input types – when in an error state, the red border is _replaced_ with a black border.

But in Firefox, you always end up with a 'triple ring' like this:

![Screenshot_2019-05-03 GOV UK Frontend(1)](https://user-images.githubusercontent.com/121939/57125732-545bf700-6d82-11e9-87f7-bb93bf23704d.png)

You also see the triple border in Chrome if you click on the 'Choose file' button (I assume because the button is focused, so this counts as 'focus within' the input.

![govuk-frontend-review-pr-1312 herokuapp com_components_file-upload_with-error-message_preview (4)](https://user-images.githubusercontent.com/121939/57125855-d0563f00-6d82-11e9-858a-2964147ecdfe.png)

Reply:
Fixed. I'm adding the same resets to `:focus-within` as to `:focus` to prevent the error styles from being applied when the focus is within.